pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py

```diff
@@ -261,41 +273,124 @@ def zombie_all_players(self, knight_list, archer_list, run):
+                upper_x_bound = min(max_x, const.SCREEN_WIDTH)
+                startx = lower_x_bound - min_x
+                starty = lower_y_bound - min_y
+                endx = 512 + upper_x_bound - max_x
+                endy = 512 + upper_y_bound - max_y
+                cropped = np.zeros_like(self.observation_spaces[agent].low)
+                cropped[startx:endx, starty:endy, :] = screen[
+                    lower_x_bound:upper_x_bound, lower_y_bound:upper_y_bound, :
+                ]
+
+            return np.swapaxes(cropped, 1, 0)
 
-        if not agent_obj.alive:
-            cropped = np.zeros((512, 512, 3), dtype=np.uint8)
         else:
-            min_x = agent_position[0] - 256
-            max_x = agent_position[0] + 256
-            min_y = agent_position[1] - 256
-            max_y = agent_position[1] + 256
-            lower_y_bound = max(min_y, 0)
-            upper_y_bound = min(max_y, const.SCREEN_HEIGHT)
-            lower_x_bound = max(min_x, 0)
-            upper_x_bound = min(max_x, const.SCREEN_WIDTH)
-            startx = lower_x_bound - min_x
-            starty = lower_y_bound - min_y
-            endx = 512 + upper_x_bound - max_x
-            endy = 512 + upper_y_bound - max_y
-            cropped = np.zeros_like(self.observation_spaces[agent].low)
-            cropped[startx:endx, starty:endy, :] = screen[
-                lower_x_bound:upper_x_bound, lower_y_bound:upper_y_bound, :
-            ]
-
-        return np.swapaxes(cropped, 1, 0)
+            # get the agent
+            i = self.agent_name_mapping[agent]
+            agent = self.agent_list[i]
+            agent_state = agent.vector_state
+
+            # get the agent position
+            agent_pos = np.expand_dims(agent_state[:2], axis=0)
+            agent_ang = np.expand_dims(agent_state[2:], axis=0)
+
+            # get vector state of everything
+            state = self.get_vector_state()
+            all_pos = state[:, :2]
+            all_ang = state[:, 2:]
+
+            # compute the angles that everything is facing
+            all_ang = np.arctan2(all_ang[:, 0], all_ang[:, 1])
+            agent_ang = np.arctan2(agent_ang[:, 0], agent_ang[:, 1])
+
+            # get relative angles that everything is facing relative
+            # to the current agent
+            rel_ang = all_ang - agent_ang
+
+            # get rotation matrix
+            c, s = np.cos(rel_ang), np.sin(rel_ang)
+            rot_mat = np.array([[c, -s], [s, c]])
+            rot_mat = np.transpose(rot_mat, (2, 0, 1))
+
+            # unit vector
+            unit_vec = np.array([0, -1])
+
+            # get rotated relative unit vectors
+            rel_ang = rot_mat @ unit_vec
+
+            # get relative positions scaled to screen size
+            rel_pos = all_pos - agent_pos
+            rel_pos = rel_pos / const.SCREEN_DIAG
+
+            # give more emphasis to closer positions
+            rel_pos = np.sign(rel_pos) * (1 - abs(rel_pos))
```

Comment:
Not sure I like this strategy. In the current map, the difference between being inside a knights attack range and outside will be like 0.98 vs 0.96, which is pretty hard for an ML algorithm to split.

Reply:
'Fixed' - the new regime uses `1 / rel_distance` for per-agent observation, and uses a linear scale in [0, 1] for global and absolute states.

Reply:
I know I suggested this approach, but some more thought is needed, there are a lot of odd observations that can occur. 

1. Some threshold should be given so that this doesn't result in infinite blowup of the observation value. Basically some sense of when a position is "equal". A more continuous way of doing this is increase the epsilon in the denominator, perhaps to 0.1, i.e. `rel_pos = 1. / (rel_pos + 1e-1)`. This will bound the observation between [-10,10] with the distance function effectively being linear if the agent is really close. There are a lot of options here, just a suggestion for one of them. 
2. If dx=100 and dy=1, then this observation strategy will yield (0.01, 1) for the observation which is a bit strange, the value in the y direction is quite large, considering how far away the object is. Perhaps rather than the coordinates being in the denominator, the full euclidian distance should be in the denominator. `rel_pos = 1. / (np.linalg.norm(rel_pos) + 1e-6)`. 
3. If the relative position is calculated in a non-linear way like this, rather than in the obvious (x-x') method, then it will be hard to determine the zombie's absolute positions with just your own absolute position and a non-linear relative positions to the zombie.
4. The ` rel_pos = np.sign(rel_pos) * (1 - abs(rel_pos))` doesn't really make sense in combination with this inverse magnitude logic.

Reply:
Some opinions regarding those above:
1. Yes, I agree with this, I've fixed it in the latest commit.
2. I think this is generally ok. It's probably pretty easy for the network to learn that the true threat of losing is actually `rel_x * rel_y`, and not just `max(rel_x, rel_y)`. This is the basis for disentangled x and y axis anyway so it should probably be ok?
3. This is definitely true too, but it's a tradeoff between giving more emphasis to nearer zombies vs. being able to have nice linear space. In any case, I am open to any other function. 
4. Removed this, forgot to remove it earlier.